src/abi.md

```diff
@@ -1,21 +1,246 @@
+> Two integer types with different signedness, such as `u8` and `i8` are not *abi compatible*.
+
+```rust
+#[cfg(target_pointer_width = "32")]
+fn foo(x: u32) -> u32 {
+    x
+}
+#[cfg(target_pointer_width = "64")]
+fn foo(x: u64) -> u64 {
+    x
+}
+
+fn main() {
+    let f: fn(usize) -> usize = unsafe { core::mem::transmute(foo as fn(_) -> _) };
+    let x = 0usize;
+    let y = f(x);
+    assert_eq!(x, y);
+}
+```
+
+r[abi.compatibility.char]
+The type [`char`] is *abi compatible* with the type [`u32`][integer types].
+
+```rust
+fn foo(x: char) -> u32{
+    x as u32
+}
+
+fn main(){
+    let f: fn(u32)->char = unsafe{core::mem::transmute(foo as fn(_)->_)};
+    let x = b'A' as u32; // ascii character indices are the same as Unicode character indices
+    let y = f(x);
+    assert_eq!(y, 'A');
+}
+```
+
+r[abi.compatibility.pointer]
+Two [pointer types], `*mut T` and `*const U`, are *abi compatible* if the *metadata type*s of `T` and `U` are the same type.
+
+> [!NOTE]
+> [`core::marker::Sized`] types have a *metadata type* of `()`.
+
+> [!NOTE]
+> With transitivity, this applies regardless of the mutability of either pointer type
+
+```rust
+unsafe fn foo(x: *mut u32){
+   unsafe{x.write(5);}
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
+    let mut val: u32 = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.reference-box]
+The types [`&T`], [`&mut T`], [`alloc::boxed::Box<T>`], and [`core::ptr::NonNull<T>`], are *abi compatible* with `*const T`
+
+> [!NOTE]
+> With transitivity, they are also *abi compatible* with each other, and with `*mut T`, as well as references/`Box` to different types that have the same *metadata type*.
+
+```rust
+fn foo(x: &mut i32){
+   *x = 5;
+}
+
+fn main(){
+    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
+    let mut val = 0;
+    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
+    unsafe{f(ptr);}
+    assert_eq!(val, 5);
+}
+```
+
+r[abi.compatibility.core]
+The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.
+
+r[abi.compatibility.transparent]
+A [`struct`] declared with the `transparent` representation is *abi compatible* with its field that does not have size 0 and alignment 1, if such a field exists.
+
+r[abi.compatibility.zst]
+Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.
+
+r[abi.compatibility.discriminant]
+If `T` is an a type listed in [layout.repr.rust.option.elision], and `U` is the type of the *elision candidate field*, then `T` is layout compatible with `U`.
```

Comment:
The referenced section here doesn't exist yet, does it?

Reply:
Oh this is just a normal markdown link for now, I see.

Reply:
Getting rid of the hard dependency on #1654.
If #1654 lands first, then I'll remove the link. If this lands before #1654 I'll make #1645 remove the link.